### ISSUE TITLE
Support half-precision (F16) weight storage in .bpk files

### ISSUE THREAD

**Issue by antimora** [tracel-ai/burn-onnx#247]
## Summary

ONNX models typically store weights in F32 (4 bytes per element). For inference-only deployment, most weights can be safely stored in F16 (2 bytes), cutting `.bpk` file size roughly in half with minimal accuracy loss. This is standard practice (PyTorch's `model.half()`, ONNX Runtime weight-only quantization, TFLite, CoreML).

Currently burn-onnx preserves original ONNX dtypes but has no option to downcast weights to F16.

## Scope

Eligible weights are stored as F16 in the `.bpk` file. Ineligible weights (BatchNorm, Constants) stay F32. Since Burn does **not** auto-cast dtypes on load, F16 weights produce F16 tensors at runtime. This creates dtype mismatches at node boundaries (e.g., Conv2d outputs F16 feeding into BatchNorm expecting F32). The solution is to insert explicit Cast nodes in the graph where mismatches occur.

- **Benefit:** ~50% smaller `.bpk` files, F16 inference on capable backends (WGPU, CUDA)
- **Trade-off:** small precision loss for F16 layers; explicit casts at mixed-precision boundaries
- **Target dtype:** F16 only (not BF16)

## User API

```rust
ModelGen::new()
    .input("model.onnx")
    .out_dir("model/")
    .half_precision(true)  // NEW: store eligible weights as F16
    .run_from_script();
```

## Which Weights Get Converted

**Converted to F16** (large weight matrices, robust to F16):

| Node Type | Weights |
|-----------|---------|
| Linear | weight, bias |
| Conv1d/2d/3d | weight, bias |
| ConvTranspose1d/2d/3d | weight, bias |
| DeformConv2d | weight, bias |
| LayerNorm | gamma, beta |
| GroupNorm | gamma, beta |
| InstanceNorm | gamma, beta |
| LSTM / GRU / RNN | all gate weights |
| PReLU | alpha |

**Kept in F32** (excluded from conversion):

| Node Type | Weights | Why |
|-----------|---------|-----|
| BatchNorm | all (gamma, beta, running_mean, running_var) | running_var can underflow in F16 (min normal ~6e-5); keeping all in F32 is simpler |
| Constant | all tensor data | Mixed purposes (indices, thresholds, shapes); blanket conversion risky |

### Key Insight: Burn Does Not Auto-Cast on Load

`burn-store`'s `Applier` loads tensors with `Tensor::from_data_dtype(data, device, snapshot.dtype)`, preserving the stored dtype. F16 in `.bpk` means F16 tensors at runtime. This means mixed-precision graphs (some F16, some F32 weights) will have dtype mismatches at node boundaries that must be resolved explicitly.

### Step 1: burn-onnx - Downcast Eligible Weights to F16

**Per-node eligibility via NodeCodegen trait:**

```rust
/// Whether this node's weights are eligible for half-precision storage.
/// Default: true. Override to exclude specific node types.
fn half_precision_eligible(&self) -> bool { true }
```

BatchNorm and Constant override to return `false`.

**Threading the flag:**

1. `ModelGen.half_precision: bool` field + `.half_precision(bool)` builder method
2. Passed to `BurnGraph` (new field)
3. `BurnGraph::with_burnpack()` reads it
4. `collect_snapshots_from_nodes()` passes to `NodeCodegen::collect_snapshots()`
5. Each node's `collect_snapshots` checks `half_precision && self.half_precision_eligible()` then passes `Some(DType::F16)` or `None` to `create_lazy_snapshot()`

**Conversion point:**

`create_lazy_snapshot()` in `node_traits.rs` gets a `target_dtype: Option<DType>` param. Inside the lazy closure, calls `data.convert_dtype(target)` when set.

### Step 2: onnx-ir - Propagate Actual Dtypes

When weights are downcast to F16, update their `ArgType` dtypes accordingly. Type inference propagates F16 through the graph (e.g., Conv2d with F16 weight produces F16 output).

### Step 3: Graph Pass - Insert Cast Nodes at Mismatch Boundaries

After type inference, walk the graph. At each node, if input dtypes don't match (e.g., one F16, one F32), insert a Cast node to upcast to the wider dtype. The Cast node already exists in onnx-ir with working codegen in burn-onnx.

This keeps individual node codegen **unchanged**. Every node sees matching dtypes because Cast nodes resolve mismatches before inputs arrive.

### Files to Modify

**burn-onnx (weight storage):**
- `crates/burn-onnx/src/model_gen.rs` - field + builder
- `crates/burn-onnx/src/burn/graph.rs` - thread through snapshot collection
- `crates/burn-onnx/src/burn/node_traits.rs` - modify `create_lazy_snapshot`, add trait default
- `crates/burn-onnx/src/burn/node/batch_norm.rs` - override `half_precision_eligible`
- `crates/burn-onnx/src/burn/node/constant.rs` - override `half_precision_eligible`
- All node files with `collect_snapshots` - accept and forward flag
- `crates/burn-onnx/src/burn/node_codegen.rs` - dispatch macro updated for new trait method

**onnx-ir (dtype propagation + cast insertion):**
- Type inference updates to propagate F16 dtypes from downcast weights
- New graph pass to detect dtype mismatches and insert Cast nodes

**CLI:**
- `onnx2burn` binary: `--half-precision` flag

### Verification

- Unit test: snapshot of a Linear node with `half_precision=true` produces F16 `TensorData`
- Unit test: Cast node inserted when Conv2d (F16 output) feeds into BatchNorm (F32 input)
- Integration test: convert a model with `half_precision`, verify `.bpk` is ~50% smaller, verify model loads and produces correct output on both F16-capable and F32-only backends

**Comment by antimora** [comment]
@AdrianEddy, @laggui Adding this feature. F16 is only a wire format for the weights file. Running model in half precision will require some more thought and testing. 

Tell your comments.

**Comment by antimora** [comment]
Nothing here depends on upstream burn; the design above is complete. It touches every node's `collect_snapshots` and adds a dtype-reconciliation pass, so it should get a nod from @laggui on the Cast-insertion approach before anyone starts.
